### PR TITLE
Answered exercises

### PR DESCRIPTION
**Pull request by tmk-a**

### CODE REVIEW

File: content/member/tamaki/chap01/1_values/README.mdx [hunk 1]
```diff
@@ -1,4 +1,62 @@
 # Study Note
+  - What is a value in JavaScript?
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/README.mdx:2]: Make this to subheading.
We can use #, ##, ###, #### to represent headings.
more # is for more sub level.
```markdown
### What is a value in Javascript?
```

File: content/member/tamaki/chap01/1_values/README.mdx [hunk 2]
```diff
@@ -1,4 +1,62 @@
...
+  A value in JavaScript is a piece of data that can be stored in a variable, returned from a function, or passed around in a program. 
+
+  - How do values represent data in your program?
+  Values in JavaScript represent data by being stored in variables and manipulated through operations. There are two main types of values:
+  Primitive values (e.g., numbers, strings, booleans) are immutable and stored directly in memory.
+  Reference values (e.g., objects, arrays) are mutable and stored as references in memory.
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/README.mdx:8]: These are list, so using bullet point will be better for readability.
We use - to make list
```markdown
###  How do values represent data in your program?
Values in JavaScript represent data by being stored in variables and manipulated through operations. There are two main types of values:
- Primitive values (e.g., numbers, strings, booleans) are immutable and stored directly in memory.
- Reference values (e.g., objects, arrays) are mutable and stored as references in memory.

File: content/member/tamaki/chap01/1_values/README.mdx [hunk 3]
```diff
@@ -1,4 +1,62 @@
...
+
+
+  - The seven primitive types in JavaScript
+  Number,String,Boolean,Undefined,Null,Symbol,BigInt
+  Symbol: Represents a unique and immutable value, often used for object properties.
+  BigInt: Represents large integers beyond Number.MAX_SAFE_INTEGER
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/README.mdx:14]: Thease also could be list
```markdown
### The seven primitive types in JavaScript
-  Number,String,Boolean,Undefined,Null,Symbol,BigInt
- Symbol: Represents a unique and immutable value, often used for object properties.
- BigInt: Represents large integers beyond Number.MAX_SAFE_INTEGER
```

File: content/member/tamaki/chap01/1_values/README.mdx [hunk 4]
```diff
@@ -1,4 +1,62 @@
...
+
+  - Characteristics of primitive types with examples (e.g., immutability).
+  let str = "hello";
+  str.toUpperCase(); // Returns "HELLO", but `str` remains "hello"
+  console.log(str); // "hello"
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/README.mdx:19]: You can put ``` to make code block in markdown!

```markdown
```javscript
let str = "hello";
str.toUpperCase(); // Returns "HELLO", but `str` remains "hello"
console.log(str); // "hello"
`` `
```

(I just put space to show you the raw ```, when you use this, don't space them)

File: content/member/tamaki/chap01/1_values/README.mdx [hunk 5]
```diff
@@ -1,4 +1,62 @@
...
+  
+  - How type coercion works in operations like addition, subtraction, and comparison.
+  ・Addition (+)
+  If one operand is a string, the other is converted to a string.
+  ・Subtraction (-)
+  Converts both operands to numbers.
+  ・Comparison (==)
+  Converts types to compare values.
 
 # Exercises
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/README.mdx:62]: Add your PR number in `- #[number]` format and each link of exercsie PR here
```markdown
# Exercises
- #12
    - [exersise01 - Type Coercion Calculator](https://github.com/navynj/JS-Dive-Study/pull/12/files#diff-c535bf28ce97af99890de22d84bbf52f36c48a726d5ef95b4fcaeca4894a4e3a)
    - [exersise02 - Game Management System](https://github.com/navynj/JS-Dive-Study/pull/12/files#diff-46807234f1b61c9cda9122956c0493d5779d5f31e46099c910191408ad147c47)
```

**Review by navynj — COMMENTED** [review]
Reference this link to get used to markdown syntax.
- https://www.markdownguide.org/basic-syntax/

You can see how your study note currently looks like on our webpage here
- https://js-dive-study-q7jxs7lr8-navynjs-projects.vercel.app/yui/chapter/1

File: content/member/tamaki/chap01/1_values/exercise01.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            answer.push(a - b);
+            break;
+        case "divide":
+            answer.push(a / b);
+            break;
+    }
+
+    if (!answer[0]) {
+        answer[0] = "Error: Invalid input: both operands must be numbers or coercible to numbers.";
+    } else if (typeof answer[0] === "number" && ((typeof a === "string" && typeof b === "number") || (typeof a === "number" && typeof b === "string") || (typeof a === "string" && typeof b === "string"))) {
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise01.js:22]: - You check `typeof a` and `typeof b` a lot. How about make these as variable?
- Condition seems quite complicated. We can improve this.
    - `(typeof a === "string" && typeof b === "number") || (typeof a === "number" && typeof b === "string")`
        - This condition is same as `typeof a` is not equal to `typeof b`
    - `typeof a === "string" && typeof b === "string"))` 
        - It has no problem, but I prefer checking `typeof answer[0]` for short code
- These only work for string or number. It's ok because instruction doesn't say about other types, but it's good to challenge for other type coercion cases.

File: content/member/tamaki/chap01/1_values/exercise01.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        if(typeof a === "string" && typeof b === "string"){
+            answer.push(`Warning: Type coercion occurred: "${a}" (string → number), "${b}" (string → number)`);
+        } else if (typeof a === "string") {
+            answer.push(`Warning: Type coercion occurred: "${a}" (string → number)`);
+        } else if (typeof b === "string") {
+            answer.push(`Warning: Type coercion occurred: "${b}" (string → number)`);
+        }
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise01.js:29]: These part have same warning format, so we can shorthand this part as well.
- All condition in this is same as wheter `typeof a` is `string` **OR** `typeof b` is `string`.
- Use variable to hold specific warning message for `a` and `b`. Since `answer[0]` is number here, if the original variable type was string, we could know type coercion occured so write message. Otherwise, message will be empty string(`''`);
    ```javascript
    let warning = 'Warning: Type coercion occurred: ';
    const warnA =  // your code here;
    const warnB =  // your code here;
    answr.push(warning + warnA + warnB);
    ```

File: content/member/tamaki/chap01/1_values/exercise01.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    } else if (typeof answer[0] === "string" && ((typeof a === "string" && typeof b === "number") || (typeof a === "number" && typeof b === "string"))) {
+        if (typeof a === "number") {
+            answer.push(`Warning: Type coercion occurred: ${a} (number → string)`);
+        } else if (typeof b === "number") {
+            answer.push(`Warning: Type coercion occurred: ${b} (number → string)`);
+        }
+        answer.push(`Note: Implicit coercion resulted in string concatenation.`);
+    }
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise01.js:37]: Based on upper review coment, you could optimize this part as well.

File: content/member/tamaki/chap01/1_values/exercise01.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    str = `Output: ${answer[0]}`;
+    if (answer.length >= 2) {
+        str += `\n${answer[1]}`;
+    }
+    if (answer.length === 3) {
+        str += `\n${answer[2]}`;
+    }
+
+    return str;
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise01.js:47]: Not neccessary, but it's better to add `\n` in the end to divide previous log.
```
return str + '\n'
```

**Review by navynj — COMMENTED** [review]
Review for **exercise01**

File: content/member/tamaki/chap01/1_values/exercise02.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
+
+function performAction(player, action, value) {
+    switch (action) {
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:3]: Using `switch` here is nice choice!

File: content/member/tamaki/chap01/1_values/exercise02.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        case "gainXP":
+            player.level += value;
+            break;
+        case "takeDamage":
+            if (player.stats.health > value) {
+                player.stats.health -= value;
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:9]: If `health` is `20` and damage value is `50`, then health should be 0, not remaining 20 or not going down to zero.

Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:9]: If `health` is `20` and damage value is `50`, then health should be 0, not remaining 20 or not going down to zero.

File: content/member/tamaki/chap01/1_values/exercise02.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    let updatePlayer = {};
+
+    switch (command) {
+        case "save":
+            updatePlayer = structuredClone(player);
+            playerHistory.push(updatePlayer);
+            break;
+        case "load":
+            updatePlayer = playerHistory[i];
+            break;
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:31]: - What if user tries to load savepoint at not existing index?
- Test this case:
    ```javascript
    const savedState = savePointManager(null, 'load', 0);
    console.log('After load 0', savePointManager(null, 'load', 0).level);
    savedState.level = 999;
    console.log('After reload 0', savePointManager(null, 'load', 0).level); // (Should remain unchanged)
    ```

File: content/member/tamaki/chap01/1_values/exercise02.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        case "rollback":
+            updatePlayer = playerHistory.pop(playerHistory);
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:33]: - You're not returning any value here.
- What if user tries to rollback when no save points available to rollback?

File: content/member/tamaki/chap01/1_values/exercise02.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+// Load the first save point
+const firstSave = savePointManager(null, "load", 0);
+console.log(firstSave.level); // Output: 10
+console.log(firstSave.stats.health); // Output: 100
+
+// Rollback to the last save point
+const lastSave = savePointManager(null, "rollback");
+console.log(lastSave.level); // Output: 15
+console.log(lastSave.stats.health); // Output: 80
```
Comment (navynj) [inline content/member/tamaki/chap01/1_values/exercise02.js:66]: Try this examples:
```javascript
const player = {
  id: 1,
  level: 10,
  inventory: ['sword', 'shield'],
  stats: { health: 100, mana: 50 },
};

// ✅ Initial save state
console.log('=== Initial Save ===');
savePointManager(player, 'save');

// ✅ Basic gameplay actions
console.log('=== Basic Gameplay Actions ===');
performAction(player, 'gainXP', 5);
console.log(player.level); // 15

performAction(player, 'takeDamage', 20);
console.log(player.stats.health); // 80

performAction(player, 'addItem', 'potion');
console.log(player.inventory); // ["sword", "shield", "potion"]

// ✅ Save state after initial actions
console.log('=== Save After First Actions ===');
savePointManager(player, 'save');

// ✅ Edge case tests
console.log('=== Edge Cases ===');
// Ensure health does not drop below 0
performAction(player, 'takeDamage', 200);
console.log(player.stats.health); // 0

// Check if duplicate items can be added
performAction(player, 'addItem', 'potion');
console.log(player.inventory); // ["sword", "shield", "potion", "potion"]

// ✅ Save state after edge cases
console.log('=== Save After Edge Cases ===');
savePointManager(player, 'save');

// ✅ Load specific save points
console.log('=== Load Specific Save Points ===');
console.log(savePointManager(null, 'load', 0)); // First save point (level 10, health 100)
console.log(savePointManager(null, 'load', 1)); // Second save point (level 15, health 80)
console.log(savePointManager(null, 'load', 2)); // Third save point (health 0, inventory with two potions)

// ✅ Rollback to last save
console.log('=== Rollback to Last Save ===');
const rolledBackState = savePointManager(null, 'rollback');
console.log(rolledBackState); // Third save point (health 0, two potions)

// ✅ Verify loading an invalid save point index (should throw an error)
console.log('=== Load Invalid Save Point Index (Expect Error) ===');
try {
  savePointManager(null, 'load', 10); // Non-existent index
} catch (error) {
  console.log(error.message); // "Invalid save point index"
}

// ✅ Immutable save test (modifying a loaded save should not affect the stored data)
console.log('=== Immutable Save Test ===');
const savedState = savePointManager(null, 'load', 0);
savedState.level = 999;
console.log(savePointManager(null, 'load', 0).level); // 10 (Should remain unchanged)

// ✅ Deep Copy validation (ensure stats are not shared by reference)
console.log('=== Deep Copy Validation ===');
performAction(player, 'takeDamage', 30);
console.log(player.stats.health); // 0 (Already 0, should not go negative)

const deepCopyTest = savePointManager(null, 'load', 0);
console.log(deepCopyTest.stats.health); // 100 (Should remain unchanged)

// ✅ Consecutive rollback tests
console.log('=== Consecutive Rollbacks ===');
savePointManager(null, 'rollback'); // Restores second save point
savePointManager(null, 'rollback'); // Restores first save point

// ✅ Verify rollback with no save points left (should throw an error)
console.log('=== Rollback Without Save Points (Expect Error) ===');
try {
  savePointManager(null, 'rollback'); // Expected error
} catch (error) {
  console.log(error.message); // "No save points available to rollback"
}
```

**Review by navynj — COMMENTED** [review]
Review for **exercise02**
